dffml/service/dev.py

```diff
@@ -756,6 +757,138 @@ async def run(self):
             self.logger.debug("Updated version file %s", version_file)
 
 
+class CommitLintError(Exception):
+    pass
+
+
+class LintCommits(CMD):
+    """
+    Enforce commit message style 
+    """
+
+    def _get_ignore_filter(self):
+        return lambda x: not x.endswith("_")
+
+    def _execute_func_list(self, param, func_list):
+        for func in func_list:
+            param = func(param)
+        return param
+
+    async def _get_file_mutations(self):
+        no_mutation = lambda x: x
+        mutation_func_factory = lambda ext: lambda x: x + ext
+        extensions = await self._get_all_exts()
+        mutations = {
+            "prefix_mutations": [
+                no_mutation,
+                lambda x: "." + x,
+                lambda x: "dffml/" + x,
+            ],
+            "suffix_mutations": [
+                no_mutation,
+                *[mutation_func_factory(ext) for ext in extensions],
+            ],
+        }
+        return mutations
+
+    async def _get_cmd_output(self, cmd: List[str]):
+        print(f"$ {' '.join(cmd)}")
+        proc = await asyncio.create_subprocess_exec(
+            *cmd, stdout=asyncio.subprocess.PIPE
+        )
+        await proc.wait()
+        stdout, _ = await proc.communicate()
+        if proc.returncode != 0:
+            raise RuntimeError(stdout)
+        output = stdout.decode().strip()
+        return output
+
+    async def _get_relevant_commits(self):
+        current_branch = await self._get_current_branch()
+        cmd = [
+            "git",
+            "log",
+            "--no-merges",
+            "--oneline",
+            "--format=%s",
+            current_branch,
+            "^master",  #! This needs to change when master is renamed to main.
+        ]
+        commits = await self._get_cmd_output(cmd)
+        commits_list = commits.split("\n")
+        return commits_list
+
+    async def _get_current_branch(self):
+        cmd = ["git", "rev-parse", "--abbrev-ref", "HEAD"]
```

Comment:
`git log -n 1 --decorate=full | head -n 1 | awk '{print $NF}' | sed -e 's/)//g'`

- `git log -n 1 --decorate=full` gives you the first commit.
- head takes first line
- awk prints the last `.split()`
- remove the last character (which is `)`)

So this should be the same as:

```python
output.split()[-1][:-1]
```

```python
>>> "commit 8684c4ef6623aad62c41e3c452957ba240bbab95 (HEAD -> commit_msg_lint)".split()[-1][:-1]
'commit_msg_lint'
```

```
commit 8684c4ef6623aad62c41e3c452957ba240bbab95 (HEAD -> commit_msg_lint)
Author: Saahil Ali <programmer290399@gmail.com>
Date:   Fri Jun 18 21:32:46 2021 +0530

    service : dev : changed current branch cmd

```

Reply:
Still returning wrong branch name 😞

Reply:
Getting name of the current branch within GitHub actions might be possible using the `$GITHUB_REF` variable , would try using that to get the name of the branch

Reply:
Was able to get the correct branch name by using the env variable `$GITHUB_HEAD_REF` , `$GITHUB_REF` would work for pushes and not PRs , 

However, This didn't solve the issue and a new error can be seen below :  
```
+ run_commit
+ dffml service dev lint commits
$ git log --no-merges --oneline --format=%s commit_msg_lint ^master
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/bin/dffml", line 33, in <module>
    sys.exit(load_entry_point('dffml', 'console_scripts', 'dffml')())
  File "/home/runner/work/dffml/dffml/dffml/util/cli/cmd.py", line 288, in main
    result = loop.run_until_complete(cls._main(*argv[1:]))
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "/home/runner/work/dffml/dffml/dffml/util/cli/cmd.py", line 238, in _main
    result = await cls.cli(*args)
  File "/home/runner/work/dffml/dffml/dffml/util/cli/cmd.py", line 224, in cli
    return await cmd.do_run()
  File "/home/runner/work/dffml/dffml/dffml/util/cli/cmd.py", line 203, in do_run
    return await self.run()
  File "/home/runner/work/dffml/dffml/dffml/service/dev.py", line 913, in run
    commits_list = await self._get_relevant_commits()
  File "/home/runner/work/dffml/dffml/dffml/service/dev.py", line 825, in _get_relevant_commits
    commits = await self._get_cmd_output(cmd)
  File "/home/runner/work/dffml/dffml/dffml/service/dev.py", line 805, in _get_cmd_output
    raise RuntimeError(stderr)
RuntimeError: b"fatal: ambiguous argument 'commit_msg_lint': unknown revision or path not in the working tree.\nUse '--' to separate paths from revisions, like this:\n'git <command> [<revision>...] -- [<file>...]'\n"
```

Reply:
would need to change this approach to use current branch without naming it , maybe that'd solve the issue

Reply:
Tried using `$ git cherry -v master`  to get commits on the current branch which are not on master, but it resulted in the following error : 
```
+ run_commit
+ dffml service dev lint commits
$ git cherry -v master
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/bin/dffml", line 33, in <module>
    sys.exit(load_entry_point('dffml', 'console_scripts', 'dffml')())
  File "/home/runner/work/dffml/dffml/dffml/util/cli/cmd.py", line 288, in main
    result = loop.run_until_complete(cls._main(*argv[1:]))
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "/home/runner/work/dffml/dffml/dffml/util/cli/cmd.py", line 238, in _main
    result = await cls.cli(*args)
  File "/home/runner/work/dffml/dffml/dffml/util/cli/cmd.py", line 224, in cli
    return await cmd.do_run()
  File "/home/runner/work/dffml/dffml/dffml/util/cli/cmd.py", line 203, in do_run
    return await self.run()
  File "/home/runner/work/dffml/dffml/dffml/service/dev.py", line 916, in run
    commits_list = await self._get_relevant_commits()
  File "/home/runner/work/dffml/dffml/dffml/service/dev.py", line 826, in _get_relevant_commits
    commits = await self._get_cmd_output(cmd)
  File "/home/runner/work/dffml/dffml/dffml/service/dev.py", line 805, in _get_cmd_output
    raise RuntimeError(stderr)
RuntimeError: b'fatal: unknown commit master\n'
```

Reply:
There's only one branch in the PR ci machine : 
```bash
$ git branch
* (HEAD detached at pull/1076/merge)
```

Reply:
Need to look into the complete workflow , to inspect and change it to fix this bug

Reply:
Solved this finally by using the command `$ git cherry -v origin/master`, it wasn't working before with just `master` because it didn't have a local copy so it has to get it from the remote origin.